File: android/src/main/java/com/margelo/nitro/abrevvareactnative/AbrevvaBle.kt
package com.margelo.nitro.abrevvareactnative

import android.Manifest
import android.content.Intent
import android.os.Build
import android.provider.Settings
import androidx.annotation.ChecksSdkIntAtLeast
import androidx.annotation.RequiresPermission
import androidx.core.net.toUri
import com.evva.xesar.abrevva.ble.BleDeviceAdvertisementData as AbrevvaBleDeviceAdvertisementData
import com.evva.xesar.abrevva.ble.BleManager as AbrevvaBleManager
import com.evva.xesar.abrevva.disengage.DisengageStatusType as AbrevvaDisengageStatusType
import com.evva.xesar.abrevva.ble.BleWriteType
import com.evva.xesar.abrevva.ble.BleDevice as AbrevvaBleDevice

import com.evva.xesar.abrevva.util.bytesToString
import com.evva.xesar.abrevva.util.stringToBytes
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.core.resolved
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.UUID

const val DEFAULT_TIMEOUT: Long = 10_000


@DoNotStrip
class AbrevvaBleImpl : HybridAbrevvaBleImplSpec() {
  private var manager: AbrevvaBleManager = requireNotNull(
    NitroModules.applicationContext?.let { AbrevvaBleManager(it) }
  ) { "Application context is required to initialize manager" }
  override fun initialize(androidNeverForLocation: Boolean?): Promise<Unit> {
    return Promise.resolved()
  }

  override fun isEnabled(): Boolean {
    return manager.isBleEnabled()
  }

  override fun isLocationEnabled(): Boolean {
    return manager.isLocationEnabled()
  }

  override fun startEnabledNotifications(callback: (Boolean) -> Unit) {
    manager.startBleEnabledNotifications(callback)
  }

  override fun stopEnabledNotifications() {
    manager.stopBleEnabledNotifications()
  }

  override fun openLocationSettings() {
    val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
    NitroModules.applicationContext!!.currentActivity!!.startActivity(intent)
  }

  override fun openBluetoothSettings() {
    val intent = Intent(Settings.ACTION_BLUETOOTH_SETTINGS)
    NitroModules.applicationContext!!.currentActivity!!.startActivity(intent)
  }

  override fun openAppSettings() {
    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
    val activity = NitroModules.applicationContext!!.currentActivity!!

    intent.data = ("package:" + activity.packageName).toUri()
    activity.startActivity(intent)
  }

  private fun convertAdvertismentData(aData: AbrevvaBleDeviceAdvertisementData?): BleDeviceAdvertisementData? {
    if (aData == null) {
      return null
    }

    val mData = aData.manufacturerData

    if (mData == null) {
      return BleDeviceAdvertisementData(
        aData.rssi.toDouble(),
        aData.isConnectable,
        null
      )
    }
    val manData = BleDeviceManufacturerData(
      mData.companyIdentifier.toDouble(),
      mData.version.toDouble(),
      getComponentType(mData.componentType.toInt()),
      mData.mainFirmwareVersionMajor.toDouble(),
      mData.mainFirmwareVersionMinor.toDouble(),
      mData.mainFirmwareVersionPatch.toDouble(),
      mData.componentHAL.toString(),
      if (mData.batteryStatus) BatteryStatus.BATTERY_FULL else BatteryStatus.BATTERY_EMPTY,
      mData.mainConstructionMode,
      mData.subConstructionMode,
      mData.isOnline,
      mData.officeModeActive,
      mData.twoFactorRequired,
      mData.officeModeActive,
      mData.identifier,
      mData.subFirmwareVersionMajor?.toDouble(),
      mData.subFirmwareVersionMinor?.toDouble(),
      mData.subFirmwareVersionPatch?.toDouble(),
      mData.subComponentIdentifier
    )

    return BleDeviceAdvertisementData(
      aData.rssi.toDouble(),
      aData.isConnectable,
      manData
    )
  }

  fun getComponentType(value: Int): ComponentType {
    return ComponentType.entries.find { it.value == value } ?: ComponentType.UNKNOWN
  }

  fun getDisengageStatusType(value: com.evva.xesar.abrevva.disengage.DisengageStatusType): DisengageStatusType {
    return when (value) {
        AbrevvaDisengageStatusType.ERROR -> DisengageStatusType.ERROR
        AbrevvaDisengageStatusType.AUTHORIZED -> DisengageStatusType.AUTHORIZED
        AbrevvaDisengageStatusType.AUTHORIZED_PERMANENT_ENGAGE -> DisengageStatusType.AUTHORIZED_PERMANENT_ENGAGE
        AbrevvaDisengageStatusType.AUTHORIZED_PERMANENT_DISENGAGE -> DisengageStatusType.AUTHORIZED_PERMANENT_DISENGAGE
        AbrevvaDisengageStatusType.AUTHORIZED_BATTERY_LOW -> DisengageStatusType.AUTHORIZED_BATTERY_LOW
        AbrevvaDisengageStatusType.AUTHORIZED_OFFLINE -> DisengageStatusType.AUTHORIZED_OFFLINE
        AbrevvaDisengageStatusType.UNAUTHORIZED -> DisengageStatusType.UNAUTHORIZED
        AbrevvaDisengageStatusType.UNAUTHORIZED_OFFLINE -> DisengageStatusType.UNAUTHORIZED_OFFLINE
        AbrevvaDisengageStatusType.SIGNAL_LOCALIZATION -> DisengageStatusType.SIGNAL_LOCALIZATION
        AbrevvaDisengageStatusType.MEDIUM_DEFECT_ONLINE -> DisengageStatusType.MEDIUM_DEFECT_ONLINE
        AbrevvaDisengageStatusType.MEDIUM_BLACKLISTED -> DisengageStatusType.MEDIUM_BLACKLISTED
        AbrevvaDisengageStatusType.UNABLE_TO_CONNECT -> DisengageStatusType.UNABLE_TO_CONNECT
        AbrevvaDisengageStatusType.UNABLE_TO_SET_NOTIFICATIONS -> DisengageStatusType.UNABLE_TO_SET_NOTIFICATIONS
        AbrevvaDisengageStatusType.UNABLE_TO_READ_CHALLENGE -> DisengageStatusType.UNABLE_TO_READ_CHALLENGE
        AbrevvaDisengageStatusType.UNABLE_TO_WRITE_MDF -> DisengageStatusType.UNABLE_TO_WRITE_MDF
        AbrevvaDisengageStatusType.ACCESS_CIPHER_ERROR -> DisengageStatusType.ACCESS_CIPHER_ERROR
        AbrevvaDisengageStatusType.BLE_ADAPTER_DISABLED -> DisengageStatusType.BLE_ADAPTER_DISABLED
        AbrevvaDisengageStatusType.UNKNOWN_DEVICE -> DisengageStatusType.UNKNOWN_DEVICE
        AbrevvaDisengageStatusType.TIMEOUT -> DisengageStatusType.TIMEOUT
        AbrevvaDisengageStatusType.UNKNOWN_STATUS_CODE -> DisengageStatusType.UNKNOWN_STATUS_CODE
    }
  }

  override fun startScan(
    onScanResult: (BleDevice) -> Unit,
    onScanStart: ((Throwable?) -> Unit)?,
    onScanStop: ((Throwable?) -> Unit)?,
    macFilter: String?,
    allowDuplicates: Boolean?,
    timeout: Double?
  ) {

    manager.startScan(
      { device: AbrevvaBleDevice ->
        val bleDevice: BleDevice = BleDevice(
          device.address,
          device.localName,
          this.convertAdvertismentData(device.advertisementData)
        )
        onScanResult(bleDevice)
      },
      { success: Boolean ->
        onScanStart?.invoke(if (success) null else Throwable("ScanStartError"))
      }, { success: Boolean ->
        onScanStop?.invoke(if (success) null else Throwable("ScanStartError"))
      }, macFilter, allowDuplicates, timeoutToLongOrDefault(timeout)
    )
  }

  override fun stopScan() {
    manager.stopScan()
  }

  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun connect(
    deviceId: String,
    onDisconnect: (String) -> Unit,
    timeout: Double?
  ): Promise<Boolean> {


    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }

    val promise = Promise<Boolean>()

    manager.connect(
      bleDevice, { success ->
        promise.resolve(success)
      },
      { success ->
        onDisconnect(deviceId)
      },
      timeoutToLongOrDefault(timeout)
    )

    return promise
  }

  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun disconnect(deviceId: String): Promise<Boolean> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }

    val promise = Promise<Boolean>()

    manager.disconnect(bleDevice) { success ->
      promise.resolve(success)
    }

    return promise
  }

  @OptIn(DelicateCoroutinesApi::class)
  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  override fun read(
    deviceId: String,
    service: String,
    characteristic: String,
    timeout: Double?
  ): Promise<String> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }

    val promise = Promise<String>()
    GlobalScope.launch {
      val data =
        bleDevice.read(
          UUID.fromString(service),
          UUID.fromString(characteristic),
          timeoutToLongOrDefault(timeout)
        )
      if (data != null) {
        promise.resolve(bytesToString(data))
      } else {
        promise.reject(Exception("read(): failed to read from device"))
      }
    }
    return promise
  }

  @OptIn(DelicateCoroutinesApi::class)
  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  override fun write(
    deviceId: String,
    service: String,
    characteristic: String,
    value: String,
    timeout: Double?
  ): Promise<Boolean> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<Boolean>()

    GlobalScope.launch {
      val success = bleDevice.write(
        UUID.fromString(service), UUID.fromString(characteristic),
        stringToBytes(value),
        BleWriteType.NO_RESPONSE,
        timeoutToLongOrDefault(timeout)
      )
      promise.resolve(success)
    }

    return promise
  }

  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun signalize(deviceId: String): Promise<Boolean> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<Boolean>()

    manager.signalize(bleDevice) { success: Boolean ->
      promise.resolve(success)
    }

    return promise
  }

  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  @Deprecated("Use disengageWithXvnResponse() instead.")
  override fun disengage(
    deviceId: String,
    mobileId: String,
    mobileDeviceKey: String,
    mobileGroupId: String,
    mobileAccessData: String,
    isPermanentRelease: Boolean
  ): Promise<DisengageStatusType> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<DisengageStatusType>()

    manager.disengage(
      bleDevice,
      mobileId,
      mobileDeviceKey,
      mobileGroupId,
      mobileAccessData,
      isPermanentRelease
    )
    { status ->
      promise.resolve(getDisengageStatusType(status))
    }

    return promise
  }

  @OptIn(ExperimentalStdlibApi::class)
  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun disengageWithXvnResponse(
    deviceId: String,
    mobileId: String,
    mobileDeviceKey: String,
    mobileGroupId: String,
    mobileAccessData: String,
    isPermanentRelease: Boolean
  ): Promise<XvnResponse> {
    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<XvnResponse>()

    manager.disengageWithXvnResponse(
      bleDevice,
      mobileId,
      mobileDeviceKey,
      mobileGroupId,
      mobileAccessData,
      isPermanentRelease
    )
    { status, xvnData ->
      promise.resolve(XvnResponse(getDisengageStatusType(status), xvnData?.toHexString()))
    }

    return promise
  }

  @OptIn(DelicateCoroutinesApi::class)
  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun startNotifications(
    deviceId: String,
    service: String,
    characteristic: String,
    timeout: Double,
    callback: (String) -> Unit
  ): Promise<Boolean> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<Boolean>()

    GlobalScope.launch {
      val success = bleDevice.setNotifications(
        UUID.fromString(service),
        UUID.fromString(characteristic),
        { data ->
          callback(bytesToString(data))
        },
        timeoutToLongOrDefault(timeout)
      )
      promise.resolve(success)
    }

    return promise
  }

  @OptIn(DelicateCoroutinesApi::class)
  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun stopNotifications(
    deviceId: String,
    service: String,
    characteristic: String
  ): Promise<Boolean> {

    val bleDevice = manager.getBleDevice(deviceId)
    if (bleDevice == null) {
      return Promise.rejected(Error("connect(): device not found"))
    }
    val promise = Promise<Boolean>()

    GlobalScope.launch {
      val success = bleDevice.stopNotifications(
        UUID.fromString(service),
        UUID.fromString(characteristic))
      promise.resolve(success)
    }

    return promise
  }
}

File: ios/AbrevvaBle.swift
import AbrevvaSDK
import CoreBluetooth
import Foundation

typealias NitroPromise<T> = NitroModules.Promise<T>
typealias AbrevvaBleMangager = AbrevvaSDK.BleManager
typealias AbrevvaBleDevice = AbrevvaSDK.BleDevice


enum AbrevvaBleError: Error {
    case Initialize(_ message: String)
    case BleManagerNotInitialized
    case Signalize(_ message: String)
    case NotAvailableOnIos(_ message: String)
    case OpenAppSettings(_ message: String)
    case GetDevice(_ message: String)
    case Read(_ message: String)
}

class AbrevvaBleImpl: HybridAbrevvaBleImplSpec {

  
  private var bleManager: AbrevvaBleMangager?
    private var bleDeviceMap = [String: AbrevvaBleDevice]()
  
  func initialize(androidNeverForLocation: Bool?) throws -> NitroModules.Promise<Void> {
    let promise = NitroPromise<Void>()
    bleManager = AbrevvaBleMangager { success, message in
      debugPrint("bleManager init", success)

      if success {
        promise.resolve()
      } else {
        promise.reject(withError: AbrevvaBleError.Initialize(message!))
      }
    }
    return promise
  }
  
  func isEnabled() throws -> Bool {
    let bleManager = try getBleManager()
    return bleManager.isBleEnabled()
  }
  
  func isLocationEnabled() throws -> Bool  {
    throw AbrevvaBleError.NotAvailableOnIos("isLocationEnabled")
  }

  func startEnabledNotifications(callback: @escaping (Bool) -> Void) throws {
    let bleManager = try getBleManager()
    bleManager.registerStateReceiver(callback)
  }
  
  func stopEnabledNotifications() throws {
    let bleManager = try getBleManager()
    bleManager.unregisterStateReceiver()
  }
  
  func openLocationSettings() throws {
    throw AbrevvaBleError.NotAvailableOnIos("openLocationSettings")
  }
  
  func openBluetoothSettings() throws {
    throw AbrevvaBleError.NotAvailableOnIos("openBluetoothSettings")
  }
  
  func openAppSettings() throws {
    guard let settingsURL = URL(string: UIApplication.openSettingsURLString)
    else {
      throw AbrevvaBleError.OpenAppSettings("cannot open app settings")
    }
    
    let promise = NitroPromise<Bool>()
    DispatchQueue.main.async {
      if UIApplication.shared.canOpenURL(settingsURL) {
        UIApplication.shared.open(
          settingsURL,
          completionHandler: { success in
            promise.resolve(withResult: success)
          }
        )
      } else {
        promise.reject(withError: AbrevvaBleError.OpenAppSettings("cannot open app settings"))
      }
    }
  }
  
  func startScan(onScanResult: @escaping (BleDevice) -> Void, onScanStart: (((any Error)?) -> Void)?, onScanStop: (((any Error)?) -> Void)?, macFilter: String?, allowDuplicates: Bool?, timeout: Double?) throws {

    let bleManager = try getBleManager()
    self.bleDeviceMap = [:]

    bleManager.startScan(
      { device in
        self.bleDeviceMap[device.getAddress()] = device
        let bleDevice = BleDevice.init(
          deviceId: device.address,
          name: device.getName(),
          advertisementData: self.convertAdvertismentData(device.advertisementData)
        )
        onScanResult(bleDevice)
      },
      onScanStart,
      onScanStop,
      macFilter,
      allowDuplicates ?? false,
      Int(timeout ?? 10_000)
    )
  }
  
  func stopScan() throws {
    let bleManager = try getBleManager()
    bleManager.stopScan()
    
  }
  
  func connect(deviceId: String, onDisconnect: @escaping (String) -> Void, timeout: Double?) throws -> NitroModules.Promise<Bool> {
    let bleManager = try getBleManager()
    let device = try getDevice(deviceId)
    
    
    let promise = NitroPromise<Bool>()
    Task {
      let success = await bleManager.connect(
        device,
        onDisconnect,
        Int(timeout ?? 10_000)
      )
      promise.resolve(withResult: success)
    }
    return promise
  }
  
  func disconnect(deviceId: String) throws -> NitroModules.Promise<Bool> {
    let bleManager = try getBleManager()
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<Bool>()
    Task {
      let success = await bleManager.disconnect(device)
      promise.resolve(withResult: success)
    }
    return promise
  }
  
  func read(deviceId: String, service: String, characteristic: String, timeout: Double?) throws -> NitroModules.Promise<String> {
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<String>()
    Task {
      let data = await device.read(CBUUID(string: service), CBUUID(string: characteristic), Int(timeout ?? 10_000))
      if data != nil {
        promise.resolve(withResult: dataToString(data!))
      } else {
        promise.reject(withError: AbrevvaBleError.Read("failed to read data"))
      }
    }
    return promise
  }
  
  func write(deviceId: String, service: String, characteristic: String, value: String, timeout: Double?) throws -> NitroModules.Promise<Bool> {
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<Bool>()
    Task {
      let success = await device.write(
        CBUUID(string: service),
        CBUUID(string: characteristic),
        stringToData(value),
        CBCharacteristicWriteType.withoutResponse,
        Int(timeout ?? 10_000)
      )
      promise.resolve(withResult: success)
    }
    return promise
  }
  
  func signalize(deviceId: String) throws -> Promise<Bool> {
    let bleManager = try getBleManager()
    
    guard let bleDevice = bleDeviceMap[deviceId] else {
      throw AbrevvaBleError.Signalize("No BLE device with id: '\(deviceId)")
    }
    
    let promise = NitroPromise<Bool>()
    Task {
      let result = await bleManager.signalize(bleDevice)
      promise.resolve(withResult: result)
    }
    return promise
  }
  
  @available(*, deprecated, message: "Use disengageWithXvnResponse() instead.")
  func disengage(deviceId: String, mobileId: String, mobileDeviceKey: String, mobileGroupId: String, mobileAccessData: String, isPermanentRelease: Bool) throws -> NitroModules.Promise<DisengageStatusType> {
    let bleManager = try getBleManager()
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<DisengageStatusType>()
    Task {
      let status = await bleManager.disengage(
        device,
        mobileId,
        mobileDeviceKey,
        mobileGroupId,
        mobileAccessData,
        isPermanentRelease
      )
      promise.resolve(withResult: DisengageStatusType.init(fromString: status.rawValue) ?? DisengageStatusType.error)
    }
    return promise
  }
  
  func disengageWithXvnResponse(deviceId: String, mobileId: String, mobileDeviceKey: String, mobileGroupId: String, mobileAccessData: String, isPermanentRelease: Bool) throws -> Promise<XvnResponse> {
    let bleManager = try getBleManager()
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<XvnResponse>()
    Task {
      let response = await bleManager.disengageWithXvnResponse(
        device,
        mobileId,
        mobileDeviceKey,
        mobileGroupId,
        mobileAccessData,
        isPermanentRelease
      )
      let xvnData = response.1?.toHexString() ?? nil
      let disengageStatusType = DisengageStatusType(fromString: response.0.rawValue) ?? DisengageStatusType.error
      
      promise.resolve(withResult: XvnResponse(disengageStatusType: disengageStatusType, xvnData: xvnData))
    }
    return promise
  }
  
  func startNotifications(deviceId: String, service: String, characteristic: String, timeout: Double, callback: @escaping (String) -> Void) throws -> NitroModules.Promise<Bool> {
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<Bool>()
    
    Task {
      let success = await device.setNotifications(
        CBUUID(string: service),
        CBUUID(string: characteristic),
        true,
        { value in
          if value != nil {
            callback(dataToString(value!))
          }
        }, Int(timeout)
      )
      promise.resolve(withResult: success)
    }
    return promise
  }
  
  func stopNotifications(deviceId: String, service: String, characteristic: String) throws -> NitroModules.Promise<Bool> {
    let device = try getDevice(deviceId)
    
    let promise = NitroPromise<Bool>()
    
    Task {
      let success = await device.setNotifications(
        CBUUID(string: service),
        CBUUID(string: characteristic),
        false,
        nil,
      )
      promise.resolve(withResult: success)
    }
    return promise
  }
  
  private func getBleManager() throws -> BleManager {
    guard let bleManager = self.bleManager else {
      throw AbrevvaBleError.BleManagerNotInitialized
    }
    return bleManager
  }
  
  private func getDevice(
    _ deviceId: String,
    checkConnection: Bool = true
  ) throws -> AbrevvaSDK.BleDevice {
    
    guard let device = bleDeviceMap[deviceId] else {
      throw AbrevvaBleError.GetDevice("No BLE device with id: '\(deviceId)")
    }
    if checkConnection {
      guard device.isConnected() else {
        throw AbrevvaBleError.GetDevice("not connected to BLE device with id: '\(deviceId)")
      }
    }
    return device
  }
  
  private func convertAdvertismentData(_ data: AbrevvaSDK.BleDeviceAdvertisementData?) -> margelo.nitro.abrevvareactnative.BleDeviceAdvertisementData? {
    guard let aData = data else  { return nil }
    
    
    guard let mData = aData.manufacturerData else { return
      BleDeviceAdvertisementData.init(
        rssi: Double(aData.rssi),
        isConnectable: aData.isConnectable,
        manufacturerData: nil,
      )
    }

    let manData = BleDeviceManufacturerData.init(
      companyIdentifier: Double(mData.companyIdentifier),
      version: Double(mData.version),
      componentType:ComponentType.init(rawValue: Int32(mData.componentType)),
      mainFirmwareVersionMajor: Double(mData.mainFirmwareVersionMajor),
      mainFirmwareVersionMinor: Double(mData.mainFirmwareVersionMinor),
      mainFirmwareVersionPatch: Double(mData.mainFirmwareVersionPatch),
      componentHAL: String(mData.componentHAL),
      batteryStatus: BatteryStatus.init(rawValue: mData.batteryStatus ? 1 : 0),
      mainConstructionMode: mData.mainConstructionMode,
      subConstructionMode: mData.subConstructionMode,
      isOnline: mData.isOnline,
      officeModeEnabled: mData.officeModeActive,
      twoFactorRequired: mData.twoFactorRequired,
      officeModeActive: mData.officeModeActive,
      identifier: mData.identifier,
      subFirmwareVersionMajor: Double(mData.subFirmwareVersionMajor ?? 0),
      subFirmwareVersionMinor: Double(mData.subFirmwareVersionMinor ?? 0),
      subFirmwareVersionPatch: Double(mData.subFirmwareVersionPatch ?? 0),
      subComponentIdentifier: mData.subComponentIdentifier,
    )
    
    return BleDeviceAdvertisementData.init(
      rssi: Double(aData.rssi),
      isConnectable: aData.isConnectable,
      manufacturerData: manData,
    )
  }
}
